### PR TITLE
MDS-3754: Preamble text in Draft permits.

### PR DESCRIPTION
**Pull request by areyeslo**
# Main

-  Add ability to edit and save changes of preamble text in a permit amendment draft.

# Other
- Insert variables from the Variable Condition Menu during edit at the position of the cursor.

# How to test
- Open a NoW application.
- Go to a Draft tab and start a new Draft.
- New Preamble Text field is shown under Preamble section:
image
- Click on Edit Preamble Text to start editing the text and save changes. 
image
- From de Condition Data Variables menu click on any variable to be inserted at the cursor position in the Preamble Text and Save
- Click on Download Draft to generate the document with changes in Preamble Text

# Notes
https://bcmines.atlassian.net/browse/MDS-3754

### CODE REVIEW

File: migrations/sql/V2022.02.10.16.20__add_preambule_text_column_to_permit_amendment_permit.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE PERMIT_AMENDMENT 
+    ADD COLUMN PREAMBLE_tEXT character varying(2000) DEFAULT '' NOT NULL;
```
Comment (mmumby) [inline migrations/sql/V2022.02.10.16.20__add_preambule_text_column_to_permit_amendment_permit.sql]: We should confirm the limit with the business and if it's necessary, 2000 seems a little low.

Comment (areyeslo) [reply migrations/sql/V2022.02.10.16.20__add_preambule_text_column_to_permit_amendment_permit.sql]: I'll confirm the limit. Currently, in the word template we had approx. 1000 chars.

File: services/core-api/app/api/mines/permits/permit/resources/permit.py [hunk 2]
```diff
@@ -125,6 +126,13 @@ def post(self, mine_guid):
         if not mine:
             raise NotFound('There was no mine found with the provided mine_guid.')
 
+        identity = NOWApplicationIdentity.find_by_mine_guid(mine.mine_guid)
+        description = None
+        if identity:
```
Comment (mmumby) [inline services/core-api/app/api/mines/permits/permit/resources/permit.py:131]: can you rename description to be more clear - "application_type_description" or something similar?

File: services/core-api/app/api/mines/permits/permit_amendment/models/permit_amendment.py [hunk 3]
```diff
@@ -37,6 +37,7 @@ class PermitAmendment(SoftDeleteMixin, AuditMixin, Base):
     description = db.Column(db.String)
     issuing_inspector_title = db.Column(db.String)
     regional_office = db.Column(db.String)
+    preamble_text = db.Column(db.String, nullable=False, server_default=FetchedValue())
```
Comment (mmumby) [inline services/core-api/app/api/mines/permits/permit_amendment/models/permit_amendment.py:41]: nullable=True, and I don't think `server_default` is necessary here

File: services/core-api/app/api/mines/permits/permit/resources/permit.py [hunk 4]
```diff
@@ -172,12 +180,18 @@ def post(self, mine_guid):
                                data.get('is_exploration'), data.get('exemption_fee_status_code'),
                                data.get('exemption_fee_status_note'))
 
+        is_generated_in_core = True if permit.permit_status_code == 'D' else False
+
+        preamble_text = get_preamble_text(
+            application_type_description) if is_generated_in_core else ''
+
```
Comment (mmumby) [inline services/core-api/app/api/mines/permits/permit/resources/permit.py:187]: else None

File: services/core-api/app/api/mines/permits/permit_amendment/resources/permit_amendment.py [hunk 5]
```diff
@@ -170,6 +178,7 @@ def post(self, mine_guid, permit_guid, permit_amendment_guid=None):
         now_application_guid = data.get('now_application_guid')
         populate_with_conditions = data.get('populate_with_conditions', True)
         is_generated_in_core = True if permit_amendment_status_code == "DFT" and populate_with_conditions else False
+        preamble_text = get_preamble_text(application_type_description) if is_generated_in_core else ''
```
Comment (mmumby) [inline services/core-api/app/api/mines/permits/permit_amendment/resources/permit_amendment.py:183]: else None

File: services/core-api/app/api/mines/response_models.py [hunk 6]
```diff
@@ -241,6 +245,7 @@ def format(self, value):
         'exemption_fee_status_note': fields.String,
         'site_properties': fields.List(fields.Nested(MINE_TYPE_MODEL)),
         'permit_prefix': fields.String,
+        'preamble_text': fields.String
     })
```
Comment (mmumby) [inline services/core-api/app/api/mines/response_models.py]: this only exists on the permit_amendment, right? not needed on the permit

File: services/core-web/src/components/Forms/permits/GeneratePermitForm.js [hunk 7]
```diff
@@ -300,6 +310,64 @@ export const GeneratePermitForm = (props) => {
...
+              </div>
+            )}
+            <div
+              style={
+                props.editingPreambleFlag ? { backgroundColor: "#f3f0f0", padding: "20px" } : {}
+              }
+            >
+              <br />
+              <br />
+              <Row gutter={32}>
```
Comment (mmumby) [inline services/core-web/src/components/Forms/permits/GeneratePermitForm.js:337]: can you move the `<br />` above the div? 

```
<>
<br />
<br /> 
 <div
              style={
                props.editingPreambleFlag ? { backgroundColor: "#f3f0f0", padding: "20px" } : {}
              }
            >
            ..... 
            
            
            </div> 
            </>
            ```

Comment (mmumby) [reply services/core-web/src/components/Forms/permits/GeneratePermitForm.js:337]: that didn't format very well

File: services/core-web/src/components/Forms/permits/GeneratePermitForm.js [hunk 8]
```diff
@@ -300,6 +310,64 @@ export const GeneratePermitForm = (props) => {
...
+                <Col xs={48} md={24}>
+                  <Field
+                    id="preamble_text"
+                    name="preamble_text"
+                    label="Preamble text"
+                    component={renderConfig.AUTO_SIZE_FIELD}
+                    disabled={!props.editingPreambleFlag}
+                    minRows={4}
+                    validate={maxLength(2000)}
+                  />
```
Comment (mmumby) [inline services/core-web/src/components/Forms/permits/GeneratePermitForm.js:347]: maxLength(4000)

File: services/core-web/src/components/Forms/permits/conditions/VariableConditionMenu.js [hunk 9]
```diff
@@ -23,9 +26,19 @@ const defaultProps = {
...
+      const newValues = `${condition} ${value.key}`;
+      return this.props.change(FORM.CONDITION_SECTION, "condition", newValues);
+    }
+
+    const preambleText = this.props.generatePermitFormValues.preamble_text
+      ? this.props.generatePermitFormValues.preamble_text
+      : "";
+    const newPreambleText = `${preambleText} ${value.key}`;
+    return this.props.change(FORM.GENERATE_PERMIT, "preamble_text", newPreambleText);
   }
```
Comment (mmumby) [inline services/core-web/src/components/Forms/permits/conditions/VariableConditionMenu.js:45]: I think we can simplify this if/else 

```
if (!isEmpty(this.props.conditionFormValues)) {
// condition logic
} else {
// preamble_text logic
}
```

OR you can pass the form, and the field string into the component and make this function generic.

File: services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js [hunk 10]
```diff
@@ -475,6 +479,33 @@ export class NOWPermitGeneration extends Component {
...
+        this.handleDraftPermit();
+        this.props.storeEditingPreambleFlag(false);
+      });
+  };
+
+  handleCancelPreambleTextEdit = () => {
+    this.props.reset(FORM.GENERATE_PERMIT);
+    this.props.storeEditingPreambleFlag(false);
+    // this.props.toggleEditMode();
+  };
```
Comment (mmumby) [inline services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js:507]: remove

File: services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js [hunk 11]
```diff
@@ -560,7 +591,10 @@ export class NOWPermitGeneration extends Component {
               <Popconfirm
                 placement="bottomRight"
                 title="You have unsaved changes. Are you sure you want to cancel?"
-                onConfirm={this.handleCancelDraftEdit}
+                onConfirm={() => {
+                  this.handleCancelDraftEdit();
+                  this.props.storeEditingPreambleFlag(false);
+                }}
```
Comment (mmumby) [inline services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js:596]: 'this.props.storeEditingPreambleFlag(false);' - This doesn't seem relevant to this button.

File: services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js [hunk 12]
```diff
@@ -572,7 +606,10 @@ export class NOWPermitGeneration extends Component {
               <Button
                 type="primary"
                 className="full-mobile"
-                onClick={this.handleSaveDraftEdit}
+                onClick={() => {
+                  this.handleSaveDraftEdit();
+                  this.props.storeEditingPreambleFlag(false);
+                }}
                 loading={this.props.submitting}
```
Comment (mmumby) [inline services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js:611]: same here.

File: services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js [hunk 13]
```diff
@@ -639,6 +676,12 @@ export class NOWPermitGeneration extends Component {
                     }
                     draftPermit={this.props.draftPermit}
                     draftPermitAmendment={this.props.draftPermitAmendment}
+                    storeEditingPreambleFlag={this.props.storeEditingPreambleFlag}
+                    handleSaveDraftEdit={this.handleSaveDraftEdit}
+                    handleSavePreamble={this.handleSavePreamble}
+                    handleCancelDraftEdit={this.handleCancelDraftEdit}
+                    handleCancelPreambleTextEdit={this.handleCancelPreambleTextEdit}
+                    handleChange={this.handleChange}
```
Comment (mmumby) [inline services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js]: Is this used anywhere?? Don't see it defined.

File: services/minespace-web/common/selectors/permitSelectors.js [hunk 14]
```diff
@@ -11,6 +11,7 @@ export const {
   getPermitConditions,
   getStandardPermitConditions,
   getEditingConditionFlag,
+  getEditingPreambleFlag,
```
Comment (mmumby) [inline services/minespace-web/common/selectors/permitSelectors.js:14]: Can you create a new test for this?

File: services/core-api/app/api/now_applications/models/now_application_document_type.py [hunk 15]
```diff
@@ -56,11 +56,18 @@ def transform_variables_to_data(now_application, draft_permit, mine, total_liabi
...
                 str(now_application.proposed_annual_maximum_tonnage)
                 if now_application.proposed_annual_maximum_tonnage is not None else ' ',
                 'issue_date':
                 format_datetime_to_string(draft_permit.issue_date),
+                'application_dated':
+                format_datetime_to_string(now_application.submitted_date),
+                'application_last_updated_date':
+                format_datetime_to_string(now_application.last_updated_date)
+                if now_application.last_updated_date else now_application.submitted_date,
                 'authorization_end_date':
```
Comment (mmumby) [inline services/core-api/app/api/now_applications/models/now_application_document_type.py:71]: I don't see these variables added to the VariableConditionMenu as options, have they been added to the FE too?

File: services/core-api/app/api/utils/helpers.py [hunk 16]
```diff
@@ -29,3 +29,16 @@ def format_datetime_to_string(date):
 
 def format_currency(value):
     return '${:,.2f}'.format(float(value or 0)) if value else '$0.00'
+
+
+def get_preamble_text(description):
+    new_line = '\n\n'
+    return f"{description} for the {{mine_name}} {{application_type}} project was filed with the Chief Permitting Officer, " \
+           f"submitted on {{application_dated}} and last updated on {{application_last_updated_date}}.  The application included a " \
+           f"plan of the proposed work system (“Mine Plan”) and a program for the protection and reclamation of the surface of " \
```
Comment (mmumby) [inline services/core-api/app/api/utils/helpers.py:38]: The variabled injected only have 1 curly brace surrounding  `{total_liability}`, should these only have 1?

Comment (mmumby) [reply services/core-api/app/api/utils/helpers.py:38]: I see that is populated in the UI/db correctly...so disregard.

File: services/core-api/app/api/mines/permits/permit_amendment/models/permit_amendment.py (hunk 3 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/permits/permit_amendment/models/permit_amendment.py:41]: Great call, it was defined like that when the solution included the preamble_text to be generated from the database which is not longer the case.

File: services/core-api/app/api/mines/response_models.py (hunk 6 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/response_models.py]: No needed, just in permit_amendment response model.

File: services/core-web/src/components/Forms/permits/GeneratePermitForm.js (hunk 8 above)
Comment (areyeslo) [reply services/core-web/src/components/Forms/permits/GeneratePermitForm.js:347]: Forgot to match with BE. Thanks :)

File: services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js (hunk 13 above)
Comment (areyeslo) [reply services/core-web/src/components/noticeOfWork/applications/permitGeneration/NOWPermitGeneration.js]: it is not used, it was a the handler used for inserting the variable at cursor position which is not longer included.

File: services/core-api/app/api/now_applications/models/now_application_document_type.py (hunk 15 above)
Comment (areyeslo) [reply services/core-api/app/api/now_applications/models/now_application_document_type.py:71]: Removing this, not need to include them.

File: services/core-api/app/api/utils/helpers.py (hunk 16 above)
Comment (areyeslo) [reply services/core-api/app/api/utils/helpers.py:38]: Just formatting for string in python

File: services/minespace-web/common/selectors/permitSelectors.js (hunk 14 above)
Comment (areyeslo) [reply services/minespace-web/common/selectors/permitSelectors.js:14]: Of course. Thanks for mentioning.

File: services/core-api/app/api/now_applications/models/now_application_document_type.py [hunk 17]
```diff
@@ -56,6 +56,8 @@ def transform_variables_to_data(now_application, draft_permit, mine, total_liabi
                 now_application.mine_name,
                 'mine_no':
                 now_application.mine_no,
+                'application_type':
+                now_application.notice_of_work_type.description,
                 'proposed_annual_maximum_tonnage':
```
Comment (mmumby) [inline services/core-api/app/api/now_applications/models/now_application_document_type.py:61]: wait...why were the variables removed? They must exist here on the BE, and on the FE in the variable condition menu.

**Review by JustinMacaulay — APPROVED** [review]
as per discussion with Meredith looks good.
